aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -423,6 +484,15 @@ inline Return Dispatcher::callWithDispatchKeySlowPath(const TypedOperatorHandle<
       } else {
         runRecordFunction(guard, op, dispatchKey);
       }
+      if (C10_UNLIKELY(guard.needsOutputs())) {
+        // Call the kernel and capture the outputs temporarily in result to pass
+        // to RecordFunction.
+        detail::ReturnValue<Return> result(
```

Comment:
I find it confusing that the constructor of `ReturnValue` does the actual call to the kernel. If I was scanning this op to try and find the actual kernel call, that's not where I would have expected it. It might be better to do the kernel call in here and then just pass the `Return` object into the `ReturnValue` and wrap/store it there.

Reply:
Yeah, this is a bit tricky. Here are some cases to consider:

- `Return -> void`: this case we have to do some special handling, one method is to use `constexpr` similar to `make_boxed_from_unboxed_functor::call`, or we could do something like `ReturnValue` with template specialization. Wish we have C++17, without it, the next one is tricky. 
- `Return -> obj&`: With `constexpr` without C++17, we'll need to deal with `guts::if_constexpr<has_outputs>` lambda. How do we get it out of the lambda call? Can we do it without a copy? Not very familiar with `guts::if_constexpr`. The simpler way seems to initialize `output_` in constructor, hence calling the kernel in constructor.

So I have been thinking about doing this a few different ways, but each has some trade offs to balance perf and usability. There might be some cool ideas I haven't thought about.